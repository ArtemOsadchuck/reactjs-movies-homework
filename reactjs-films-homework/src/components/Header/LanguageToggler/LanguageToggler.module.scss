.dropdownWrapper {
  align-items: flex-end;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  justify-content: center;
  position: relative;
  width: 4rem;

  @media screen and (max-width: 760px) {
    flex-grow: 0;
  }

  .dropdownBtn {
    background-color: #fff;
    border: 0;
    border-radius: 3px;
    font-weight: 600;
    height: 2rem;
    margin-right: 0.5rem;
    width: 3rem;
  }
}

.modal {
  align-items: center;
  display: flex;
  flex-direction: column;
  height: 4rem;
  justify-content: space-around;
  margin-top: 7.8rem;
  position: fixed;
  width: 4rem;
  z-index: 33;

  &:hover {
    cursor: pointer;
  }

  .modalButton {
    align-items: center;
    background-color: #fff;
    display: flex;
    height: 100%;
    justify-content: center;
    padding-top: 5px;
    width: 100%;

    &:hover {
      background-color: rgb(250, 250, 250);
      font-weight: 600;
    }
  }

  .arrow {
    border-color: transparent transparent #fff;
    border-style: solid;
    border-width: 0 10px 10px;
    height: 0;
    width: 0;
  }
}
