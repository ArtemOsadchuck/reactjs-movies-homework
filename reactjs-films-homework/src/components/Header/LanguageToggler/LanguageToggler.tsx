import React, { useState } from 'react';
import styles from './LanguageToggler.module.scss';
import { setLang } from '../../../store/rootStore/mainStore/mainSlice';
import { useAppSelector, useAppDispatch } from '../../../hooks/hooks';
import getMainData from '../../../store/rootStore/mainStore/getMainData';
import getMovieDetailsData from '../../../store/rootStore/movieDetailsPageStore/getMoviePageData/getMovieDetailsData';
import getTopBilletCastData from '../../../store/rootStore/movieDetailsPageStore/getMoviePageData/getTopBilletCastData';

const LanguageToggler: React.FC = () => {
  const [isDropDownShow, setIsDropDownShow] = useState(false);
  const appLang = useAppSelector((state) => state.mainReducer.lang);
  const dispatch = useAppDispatch();

  const mainState = {
    page: useAppSelector((state) => state.mainReducer.page),
    category: useAppSelector((state) => state.mainReducer.category),
    query: useAppSelector((state) => state.mainReducer.query),
  };
  const moviePageState = {
    lang: appLang,
    movie_id: useAppSelector((state) => state.movieDetailsReducer.movie_id),
  };
  const langArr = ['RU', 'EN'];

  const downShow = () => {
    setIsDropDownShow(() => !isDropDownShow);
  };
  const changeLang = (lang: string) => {
    const dd = { ...mainState, lang: `${lang.toLowerCase()}-${lang}` };
    dispatch(setLang(lang));
    setIsDropDownShow(() => !isDropDownShow);
    dispatch(getMainData(dd));
    dispatch(getMovieDetailsData(moviePageState));
    dispatch(getTopBilletCastData(moviePageState));
  };

  return (
    <div className={styles.dropdownWrapper}>
      <button onClick={downShow} className={styles.dropdownBtn}>
        {appLang}
      </button>
      {isDropDownShow ? (
        <>
          <div className={styles.modal}>
            <div className={styles.arrow}></div>
            {langArr.map((e) => {
              return (
                <div
                  className={styles.modalButton}
                  key={e}
                  onClick={() => changeLang(e)}
                >
                  {e}
                </div>
              );
            })}
          </div>
        </>
      ) : null}
    </div>
  );
};

export default LanguageToggler;
