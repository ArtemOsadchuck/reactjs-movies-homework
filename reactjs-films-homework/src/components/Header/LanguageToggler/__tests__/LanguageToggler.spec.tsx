import React from 'react';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';

import LanguageToggler from '../LanguageToggler';

import StoreWrapper from '../../../../__testsUtils__/storeHoc';

const renderComponent = () => {
  const { asFragment } = render(
    <StoreWrapper>
      <LanguageToggler />
    </StoreWrapper>
  );
  return asFragment();
};

describe('LanguageToggler', () => {
  let fragment: any;
  beforeEach(() => (fragment = renderComponent()));
  afterEach(() => cleanup());

  it('should renders correctly', () => {
    expect(fragment).toMatchSnapshot();
  });

  it('must be clickable', () => {
    const changeLang = () => {
      fireEvent.click(btnStart);

      const btnRU = screen.getByText(/ru/i);

      fireEvent.click(btnRU, {
        target: target,
      });
    };
    const btnStart = screen.getByText(/en|ru/i);
    const target = <div>RU</div>;

    changeLang();

    const btnAfterChange = screen.getByText(/ru/i);

    expect(btnAfterChange).toBeInTheDocument();
  });
});
