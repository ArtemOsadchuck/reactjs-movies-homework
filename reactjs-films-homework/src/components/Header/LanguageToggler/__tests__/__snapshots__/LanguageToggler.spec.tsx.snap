// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageToggler should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dropdownWrapper"
  >
    <button
      class="dropdownBtn"
    >
      EN
    </button>
  </div>
</DocumentFragment>
`;
